### PR TITLE
Redact unsafe URLs in the Trace2 output

### PR DESCRIPTION
**Pull request by dscho**
The Trace2 output can contain secrets when a user issues a Git command with sensitive information in the command-line. A typical (if highly discouraged) example is: `git clone https://user:password@host.com/`.

With this PR, the Trace2 output redacts passwords in such URLs by default.

This series also includes a commit to temporarily disable leak checking on t0210,t0211 because the tests uncover other unrelated bugs in Git.

These patches were integrated into Microsoft's fork of Git, as https://github.com/microsoft/git/pull/616, and have been cooking there ever since.

cc: Elijah Newren <newren@gmail.com>
cc: Jeff King <peff@peff.net>

### CODE REVIEW

**Comment by jeffhostetler** [comment]
I killed the osx-clang after 2h35m and restarted it. I assumed it was stuck. The osx-gcc took only 25 minutes.
I also restarted the CI / sparse jobs. It failed on a download of something.

**Comment by dscho** [comment]
@jeffhostetler do you have any suggestions how these patches (or the cover letter) could be improved before I send this off?

File: t/t0211-trace2-perf.sh [hunk 1]
```diff
@@ -268,4 +268,23 @@ test_expect_success PTHREADS 'global counter test/test2' '
...
+		git clone https://user:pwd@example.com/ clone &&
+	! grep user:pwd trace.perf &&
+
+	GIT_TRACE2_REDACT=0 GIT_TRACE2_PERF="$(pwd)/unredacted.perf" \
+		git clone https://user:pwd@example.com/ clone2 &&
+	perl "$TEST_DIRECTORY/t0211/scrub_perf.perl" <unredacted.perf >actual &&
+	grep "d0|main|start|.* clone https://user:pwd@example.com" actual &&
+	grep "d0|main|def_param|.*|remote.origin.url:https://user:pwd@example.com" actual
+'
+
```
Comment (jeffhostetler) [inline t/t0211-trace2-perf.sh:289]: _(Since I can't comment on the commit message, I'll do it here.)_

This commit just translates your t0210 test into t0211.  Should we expand the commit message to say that or squash this into the first commit?

File: trace2.h [hunk 2]
```diff
@@ -337,8 +337,8 @@ struct key_value_info;
 void trace2_def_param_fl(const char *file, int line, const char *param,
 			 const char *value, const struct key_value_info *kvi);
 
-#define trace2_def_param(param, value) \
-	trace2_def_param_fl(__FILE__, __LINE__, (param), (value))
+#define trace2_def_param(param, value, kvi) \
+	trace2_def_param_fl(__FILE__, __LINE__, (param), (value), (kvi))
```
Comment (jeffhostetler) [inline trace2.h:341]: This commit is unrelated to the radacting theme of the series, should we put it first to better isolate it?

Comment (dscho) [reply trace2.h:341]: Sounds good!

File: t/t0211-trace2-perf.sh (hunk 1 above)
Comment (dscho) [reply t/t0211-trace2-perf.sh:289]: Will do!

**Comment by dscho** [comment]
@jeffhostetler I force-pushed changes according to your feedback. Here is the range-diff:



* 3:  b8071e2c3a4 = 1:  97d17c22ff3 trace2: fix signature of trace2_def_param() macro
* 1:  5f14510c97c = 2:  0fb2ac82e3c trace2: redact passwords from https:// URLs by default
* 2:  092dd801f62 ! 3:  ea30fa3ef1b t0211: test URL redacting in PERF format

   ``````diff
   @@ Metadata
     ## Commit message ##
        t0211: test URL redacting in PERF format
    
   +    This transmogrifies the test case that was just added to t0210, to also
   +    cover the `GIT_TRACE2_PERF` backend.
   +
        Signed-off-by: Jeff Hostetler <jeffhostetler@github.com>
    
     ## t/t0211-trace2-perf.sh ##
   ``````

* 4:  56f417da1ae ! 4:  185b4de03a5 t0212: test URL redacting in EVENT format

   ``````diff
   @@ Metadata
     ## Commit message ##
        t0212: test URL redacting in EVENT format
    
   +    This transmogrifies the test case that was just added to t0210, and then
   +    transmogrified and added to t0211, to also cover the `GIT_TRACE2_EVENT`
   +    backend.
   +
        Signed-off-by: Jeff Hostetler <jeffhostetler@github.com>
    
     ## t/helper/test-trace2.c ##
   ``````

* 5:  0d2869253e1 = 5:  b7d5f133d28 t0210,t0211: unset SANITIZE_LEAK

File: t/t0212-trace2-event.sh [hunk 3]
```diff
@@ -323,4 +323,44 @@ test_expect_success 'discard traces when there are too many files' '
 	head -n2 trace_target_dir/git-trace2-discard | tail -n1 | grep \"event\":\"too_many_files\"
 '
 
+# In the following "...redact..." tests, skip testing the GIT_TRACE2_REDACT=0
+# case because we would need to exactly model the full JSON event stream like
+# we did in the basic tests above and I do not think it is worth it.
+
```
Comment (jeffhostetler) [inline t/t0212-trace2-event.sh:329]:
```
# Furthermore, the Trace2 routines print the same content in normal, perf, or event format, and in t0210 and t0211 we already tested the basic functionality, so no need to repeat it here.
#
# In this test, we use the test-helper to unit test each of the event messages where URLs can appear and confirm that they are redacted in each event.
```

Maybe that goes in the commit message ??

**Comment by jeffhostetler** [comment]
_GH is being weird as I review this, so I'll copy it here to avoid losing it._

In my t0212 commit:

```
# Furthermore, the Trace2 routines print the same content in normal, perf, or event format, and in t0210 and t0211 we already tested the basic functionality, so no need to repeat it here.
#
# In this test, we use the test-helper to unit test each of the event messages where URLs can appear and confirm that they are redacted in each event.
```

Maybe that goes in the commit message ??

**Comment by jeffhostetler** [comment]
Does the SANITIZE_LEAK commit at the end cause problems for later bisectors ??  Should it be moved up ?

**Review by jeffhostetler — APPROVED** [review]
_I got in some weird GH review state while typing while when refreshes appeared, so I'm going to submit this as is to avoid losing state._

**Comment by jeffhostetler** [comment]
the transmogrifies message doesn't fit on t0212 -- let's go with my other suggestion.

**Comment by dscho** [comment]
I force-pushed and this is the range-diff:



* 1:  0fb2ac82e3c ! 1:  0c4954bc1b0 trace2: redact passwords from https:// URLs by default

   ``````diff
   @@ Commit message
        via `GIT_TRACE_REDACT` (defaulting to `true`), we guard the new code via
        `GIT_TRACE2_REDACT` (also defaulting to `true`).
    
   +    Turn off TEST_PASSES_SANITIZE_LEAK in t0210 and t0211 tests.
   +
   +    The tests this commit uncover leaks in `builtin/clone.c` and `remote.c`.
   +    Therefore we need to turn off `TEST_PASSES_SANITIZE_LEAK`. The reasons:
   +
   +    - We observed that `the_repository->remote_status` is not released
   +      properly.
   +
   +    - We are using `url...insteadOf` and that runs into a code path where an
   +      allocated URL is replaced with another URL, and the original URL is
   +      never released.
   +
   +    - `remote_states` contains plenty of `struct remote`s whose refspecs
   +      seem to be usually allocated by never released.
   +
   +    More investigation is needed here to identify the exact cause and
   +    proper fixes for these leaks/bugs.
   +
   +    Co-authored-by: Jeff Hostetler <jeffhostetler@github.com>
   +    Signed-off-by: Jeff Hostetler <jeffhostetler@github.com>
        Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
    
     ## t/t0210-trace2-normal.sh ##
   +@@
   + 
   + test_description='test trace2 facility (normal target)'
   + 
   +-TEST_PASSES_SANITIZE_LEAK=true
   ++TEST_PASSES_SANITIZE_LEAK=false
   + . ./test-lib.sh
   + 
   + # Turn off any inherited trace2 settings for this test.
    @@ t/t0210-trace2-normal.sh: test_expect_success 'using global config with include' '
     	test_cmp expect actual
     '
   ``````

* 2:  ea30fa3ef1b ! 2:  fe87e87351a t0211: test URL redacting in PERF format

   ``````diff
   @@ Commit message
        This transmogrifies the test case that was just added to t0210, to also
        cover the `GIT_TRACE2_PERF` backend.
    
   +    Just like t0211, we now have to toggle the `TEST_PASSES_SANITIZE_LEAK`
   +    annotation.
   +
        Signed-off-by: Jeff Hostetler <jeffhostetler@github.com>
    
     ## t/t0211-trace2-perf.sh ##
   +@@
   + 
   + test_description='test trace2 facility (perf target)'
   + 
   +-TEST_PASSES_SANITIZE_LEAK=true
   ++TEST_PASSES_SANITIZE_LEAK=false
   + . ./test-lib.sh
   + 
   + # Turn off any inherited trace2 settings for this test.
    @@ t/t0211-trace2-perf.sh: test_expect_success PTHREADS 'global counter test/test2' '
     	have_counter_event "main" "counter" "test" "test2" 60 actual
     '
   ``````

* 3:  185b4de03a5 ! 3:  cea7cd28777 t0212: test URL redacting in EVENT format

   ``````diff
   @@ Metadata
     ## Commit message ##
        t0212: test URL redacting in EVENT format
    
   -    This transmogrifies the test case that was just added to t0210, and then
   -    transmogrified and added to t0211, to also cover the `GIT_TRACE2_EVENT`
   -    backend.
   +    Furthermore, the Trace2 routines print the same content in normal, perf,
   +    or event format, and in t0210 and t0211 we already tested the basic
   +    functionality, so no need to repeat it here.
   +
   +    In this test, we use the test-helper to unit test each of the event
   +    messages where URLs can appear and confirm that they are redacted in
   +    each event.
    
        Signed-off-by: Jeff Hostetler <jeffhostetler@github.com>
    
   ``````

* 4:  b7d5f133d28 < -:  ----------- t0210,t0211: unset SANITIZE_LEAK

But I guess I misunderstood and the last commit needs to move the code comment into the commit message?

File: t/t0212-trace2-event.sh (hunk 3 above)
Comment (jeffhostetler) [reply t/t0212-trace2-event.sh:329]: sorry to be confusing. my brain was a bit scattered.  let me see what you have below.

File: t/t0210-trace2-normal.sh [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 test_description='test trace2 facility (normal target)'
 
-TEST_PASSES_SANITIZE_LEAK=true
+TEST_PASSES_SANITIZE_LEAK=false
 . ./test-lib.sh
```
Comment (jeffhostetler) [inline t/t0210-trace2-normal.sh:6]: A comment on the commit message:

s/The tests this commit/The new tests added in this commit/

Maybe drop the "Turn off TEST... in t0210 and t0211 tests." sentence, since it is explained in the next paragraph (and we only modified t0210 in this commit).

File: t/t0212-trace2-event.sh (hunk 3 above)
Comment (jeffhostetler) [reply t/t0212-trace2-event.sh:329]: Let's leave the code comment here as it is.  And then make the commit message something like:

```
t0212: test URL redacting in EVENT format

Test the redact URL changes in the Trace2 EVENT format.

In the previous 2 commits, we tested basic URL redacting functionality
in NORMAL and PERF formats in t0210 and t0211.  Since the Trace2
routines print the same content regardless of format, we do not need
to repeat those tests here.

In this test, we use the test-helper to unit test each of the event
messages where URLs can appear and confirm that they are redacted in
each event.  This allows us to test redaction for events that are difficult
to setup tests for, such as child events.
```

**Comment by dscho** [comment]
Next attempt. <details><summary>Range-diff</summary>

* 1:  0c4954bc1b0 ! 1:  a1686ab52f1 trace2: redact passwords from https:// URLs by default

   ``````diff
   @@ Commit message
        via `GIT_TRACE_REDACT` (defaulting to `true`), we guard the new code via
        `GIT_TRACE2_REDACT` (also defaulting to `true`).
    
   -    Turn off TEST_PASSES_SANITIZE_LEAK in t0210 and t0211 tests.
   -
   -    The tests this commit uncover leaks in `builtin/clone.c` and `remote.c`.
   -    Therefore we need to turn off `TEST_PASSES_SANITIZE_LEAK`. The reasons:
   +    The new tests added in this commit uncover leaks in `builtin/clone.c`
   +    and `remote.c`. Therefore we need to turn off
   +    `TEST_PASSES_SANITIZE_LEAK`. The reasons:
    
        - We observed that `the_repository->remote_status` is not released
          properly.
   ``````

* 2:  fe87e87351a = 2:  e50160fedc0 t0211: test URL redacting in PERF format
* 3:  cea7cd28777 ! 3:  3e436f7a28d t0212: test URL redacting in EVENT format

   ``````diff
   @@ Metadata
     ## Commit message ##
        t0212: test URL redacting in EVENT format
    
   +    In the added tests cases, skip testing the `GIT_TRACE2_REDACT=0` case
   +    because we would need to exactly model the full JSON event stream like
   +    we did in the preceding basic tests and I do not think it is worth it.
   +
        Furthermore, the Trace2 routines print the same content in normal, perf,
        or event format, and in t0210 and t0211 we already tested the basic
        functionality, so no need to repeat it here.
   @@ t/t0212-trace2-event.sh: test_expect_success 'discard traces when there are too
     	head -n2 trace_target_dir/git-trace2-discard | tail -n1 | grep \"event\":\"too_many_files\"
     '
     
   -+# In the following "...redact..." tests, skip testing the GIT_TRACE2_REDACT=0
   -+# case because we would need to exactly model the full JSON event stream like
   -+# we did in the basic tests above and I do not think it is worth it.
   -+
    +test_expect_success 'unsafe URLs are redacted by default in cmd_start events' '
    +	test_when_finished \
    +		"rm -r trace.event" &&
   ``````


</details>

File: t/t0212-trace2-event.sh (hunk 3 above)
Comment (jeffhostetler) [reply t/t0212-trace2-event.sh:329]: no problem. the comment is not really needed with the improved commit message.

Comment (jeffhostetler) [reply t/t0212-trace2-event.sh:329]: no problem. the comment is not really needed with the improved commit message.
